File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def test(root):
            if not root: return [True, 0]
            
            left, right = test(root.left), test(root.right)
            
            height_dif = abs(left[1]-right[1])
            balanced = left[0] and right[0] and height_dif<=1 
            
            height_of_the_root = max(left[1], right[1])
            return [balanced, 1 + height_of_the_root] 
        return test(root)[0]

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int isBal(TreeNode root){
      if(root==null)return 0;
      
      int left_tree=isBal(root.left);
      
      int right_tree=isBal(root.right);
      
      if(left_tree==-1 || right_tree==-1)return -1;
      
      if(Math.abs(right_tree-left_tree)>1)return -1;
       System.out.println(Math.max(left_tree,right_tree));
      return Math.max(left_tree,right_tree)+1;
      
      
    }
  
  
    public boolean isBalanced(TreeNode root) {
        if(root==null)return true;
        return isBal(root) !=-1 ? true: false;
    }
}